### ISSUE TITLE
Evaluation returns `eval_loss = nan` when using context parallel

### ISSUE THREAD

**Issue by mingkaid** [axolotl-ai-cloud/axolotl#3026]
### Please check that this issue hasn't been reported before.

- [x] I searched previous [Bug Reports](https://github.com/axolotl-ai-cloud/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

When using context parallel, the evaluation should run as usual. Below is an example when using `fft + liger + fsdp`

```
{'eval_loss': 0.37890625, 'eval_runtime': 1.0652, 'eval_samples_per_second': 0.939, 'eval_steps_per_second': 0.939, 'epoch': 10.0}
{'loss': 0.4148, 'grad_norm': 0.482421875, 'learning_rate': 1.339745962155613e-06, 'epoch': 11.0}
```

### Current behaviour

When using context parallel > 1, the evaluation always returns `eval_loss: nan`, which makes it difficult to sweep parameters and do model selection. Below is an example when doing `fft + liger + fsdp + cp=2`

```
{'eval_loss': nan, 'eval_runtime': 5.3319, 'eval_samples_per_second': 1.313, 'eval_steps_per_second': 0.188, 'epoch': 9.0}
{'loss': 0.3877, 'grad_norm': 0.341796875, 'learning_rate': 4.894348370484648e-07, 'epoch': 9.5}
```

### Steps to reproduce

1. Install the library following instruction
2. Take the fft-liger-config, and add the suggested parameters for context parallel
3. Run `axolotl train path/to/config.yaml`

### Config yaml

```yaml
base_model: Qwen/Qwen3-8B

plugins:
  - axolotl.integrations.liger.LigerPlugin
liger_rope: true
liger_rms_norm: true
liger_glu_activation: true
liger_fused_linear_cross_entropy: true


chat_template: qwen3
datasets:
  - path: /path/to/data.parquet
    ds_type: parquet
    type: chat_template
    field_messages: messages
    message_property_mappings:
      role: role
      content: content

dataset_prepared_path: last_run_prepared
val_set_size: 0.10
output_dir: ./output/dir

sequence_len: 32768
sample_packing: false
pad_to_sequence_len: false
eval_sample_packing: false
dataloader_drop_last: false

wandb_project: 
wandb_entity: 
wandb_watch: 
wandb_name: 
wandb_log_model:

# The number of GPUs to shard the model parameters across (FSDP dimension).
dp_shard_size: 2

# Set to a divisor (> 1) of the number of GPUs available
context_parallel_size: 2  # Split sequences across 4 GPUs
# Optional; strides across the key dimension. Larger values use more memory but should make training faster.
heads_k_stride: 1
# Optional; one of "varlen_llama3" or "batch_ring". Defaults to
# "varlen_llama3" when `sample_packing: true`, and "batch_ring" otherwise.
ring_attn_func:

# macro batch size: 32 with 8 GPUs
# micro_batch_size * gradient_accumulation_steps * num_gpus / context_parallel_size = macro_batch_size
# 1 * 8 * 8 / 2 = 32
gradient_accumulation_steps: 8
micro_batch_size: 1
# num_epochs: 1
max_steps: 20
optimizer: adamw_torch_fused
lr_scheduler: cosine
learning_rate: 2e-5

bf16: auto
tf32: true

# gradient_checkpointing: true
# gradient_checkpointing_kwargs:
#   use_reentrant: false
resume_from_checkpoint:
logging_steps: 1
flash_attention: true

warmup_ratio: 0.1
# evals_per_epoch: 2
# saves_per_epoch: 1
eval_steps: 0.1
# save_steps: 1.0
save_strategy: best
metric_for_best_model: loss
greater_is_better: false
weight_decay: 0.0

fsdp_version: 2
fsdp_config:
  offload_params: false
  cpu_ram_efficient_loading: true
  auto_wrap_policy: TRANSFORMER_BASED_WRAP
  transformer_layer_cls_to_wrap: Qwen3DecoderLayer
  state_dict_type: FULL_STATE_DICT
  reshard_after_forward: true
  activation_checkpointing: false

special_tokens:
# save_first_step: true  # uncomment this to validate checkpoint saving works with your config
```

### Possible solution

Perhaps somehow the evaluation code didn't process the sequence parallel context correctly?

https://github.com/axolotl-ai-cloud/axolotl/blob/d09290f2f4d133b7bc25e015edc7261b8d14c89b/src/axolotl/train.py#L208

### Which Operating Systems are you using?

- [x] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.12

### axolotl branch-commit

main

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this bug has not been reported yet.
- [x] I am using the latest version of axolotl.
- [x] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by mingkaid** [comment]
@djsaunde following up from Discord

**Comment by NanoCode012** [comment]
> the evaluation always returns eval_loss: nan

Does it happen at every eval step or only after specific point?

**Comment by mingkaid** [comment]
> > the evaluation always returns eval_loss: nan
> 
> Does it happen at every eval step or only after specific point?

Every step, from the very beginning

**Comment by djsaunde** [comment]
Looking today, thanks for the bug report.

**Comment by djsaunde** [comment]
So, quick note on my findings so far. This NaN issue can happen in the CP setting when a sequence (here, split by CP) is completely masked. For example:

```
(Pdb) inputs
{'input_ids': tensor([[ 38214,    374,    458,   7600,    429,  16555,    264,   3383,     13,
           9645,    264,   2033,    429,  34901,  44595,    279,   1681,    382,
          14374,  29051,    510,  79235,    279,   2701,   4244,    311,   1281,
            264,  22414,  11652,     25,  41930,     11,  27796,     11,  15498,
             11,  34615,    382,  14374,   5949,    510,   1336,    665,   1330,
          15498,    311,  34615,     13, 151645, 151643, 151643, 151643, 151643,
         151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643, 151643,
         151643]], device='cuda:0'), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], device='cuda:0'), 'labels': tensor([[  -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,
           -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,
           -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,
           -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,
           -100,   -100,   -100,   -100,   -100,   -100,   1336,    665,   1330,
          15498,    311,  34615,     13, 151645,   -100,   -100,   -100,   -100,
           -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,   -100,
           -100]], device='cuda:0')}

(Pdb) inputs["attention_mask"].chunk(2, dim=1)[0]
tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
         1, 1, 1, 1, 1, 1, 1, 1]], device='cuda:0')
```

Computing the loss on this slice of the input will give NaN. In validation, this NaN gets pulled into the aggregate.

The fix would be to use nan-aware aggregations. I believe HF trainer doesn't do this because they don't (yet) support context parallelism. I can patch this in.
